buildsys: use clap for env vars

webern opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Related to, refactor before #70 

**Description of changes:**

Required input in the form of environment variables was sprinkled throughout the code. Here we aggregate the inputs into a CLI interface using Clap, while still allowing all to specified via environment variables.

**Testing done:**

- [x] Built an aws-k8s-1.27 image
- [x] Built it again and observed that packages were not rebuilt
- [x] touched a file (binutils.spec) and built it again - observed that binutils was rebuilt but other packages were not
- [x] Built an aws-ecs-2 image, many (but not all) packages were rebuilt due to changed variant
- [x] Built it again and observed that packages were not rebuilt
- [x] touched a file (binutils.spec) and built it again - observed that binutils was rebuilt but other packages were not

Also, during the refactoring process, I captured the output of the docker build command before and after this PR's changes. I found that I was getting the same build args and secrets args before and after (with the exception that SDK, TOOLCHAIN, TOKEN and NOCACHE have been added during build-variant, which seems fine).

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

bcressey commented on tools/buildsys/src/builder.rs:
```diff
@@ -222,17 +231,24 @@ impl VariantBuilder {
+            &build_variant_args.common.root_dir,
+            &build_variant_args.common.state_dir,
+            &build_variant_args.common.sdk_image,
+            &build_variant_args.common.toolchain,
```
I see there's a TODO for this, but it would be nice to have a `BuildContext` type to hold these values so that the `build` function signature isn't so complicated.

bcressey commented on tools/buildsys/src/builder.rs:
```diff
@@ -147,18 +158,16 @@ pub(crate) struct VariantBuilder;
-            .context(error::UnsupportedArchSnafu { arch: &arch })?
+        let arch = &build_variant_args.common.arch;
+        let goarch = serde_plain::from_str::<SupportedArch>(arch)
+            .context(error::UnsupportedArchSnafu { arch })?
```
Can we do this conversion from `&str` to `SupportedArch` just once, perhaps when parsing arguments? Otherwise we end up doing it in a few different places - here and every time `supported_arch` is called.

bcressey commented on tools/buildsys/src/cache.rs:
```diff
@@ -79,11 +78,9 @@ impl LookasideCache {
     /// Retrieves a file from the specified URL and write it to the given path,
     /// then verifies the contents against the SHA-512 hash provided.
-    fn fetch_file<P: AsRef<Path>>(url: &str, path: P, hash: &str) -> Result<()> {
+    fn fetch_file<P: AsRef<Path>>(url: &str, path: P, hash: &str, version: &str) -> Result<()> {
```
I don't love this function signature change - it feels like one thing to add additional context to a request header based on an environment variable, and another to take an otherwise useless parameter just to do the same.

I'd prefer to either not emit the version at all - do we really care? - or else to add it as a field to `LookasideCache`, pass it into `LookasideCache::new`, and retrieve it from `self`.

webern replied:
I don't know if we care about emitting the version, so I've kept it but with your suggestion to make it self.version.

bcressey commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,143 @@
+            "BUILDSYS_ARCH",
+            "BUILDSYS_OUTPUT_DIR",
+            "BUILDSYS_ROOT_DIR",
+            "BUILDSYS_VERSION_FULL",
```
This would trigger rebuilds of everything whenever the project version changes - new commit or dirty tree.

(The different `getenv` functions sometimes had the "rerun-if-changed" side effect and sometimes did not - it should just be the lookups in `builder.rs` that had the side effect.)

bcressey commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,143 @@
+    pub(crate) fn rerun_for_envs() {
+        Common::rerun_for_envs();
+        [
+            "BUILDSYS_PACKAGES_DIR",
+            "BUILDSYS_SOURCES_DIR",
```
I'm OK with unconditionally emitting these for variables we expect to be constant for the life of a given workspace, if we want to simplify slightly and confine the package- and variant-specific `rerun_for_envs()` to truly special cases.

bcressey commented on tools/buildsys/src/builder.rs:
```diff
@@ -590,14 +607,6 @@ where
-/// and ensure that we track it for changes, since it will directly affect the
-/// output.
-fn getenv(var: &str) -> Result<String> {
-    println!("cargo:rerun-if-env-changed={}", var);
```
It's really just this one `getenv` function that emits `rerun-if-changed`, which is why you picked up some new variables for tracking.

bcressey commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,143 @@
+            "BUILDSYS_OUTPUT_DIR",
+            "BUILDSYS_ROOT_DIR",
+            "BUILDSYS_VERSION_FULL",
+            "CARGO_MANIFEST_DIR",
```
I don't think we were emitting rerun-if-changed for this previously under any condition.
```suggestion
```

bcressey commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,143 @@
+        [
+            "BUILDSYS_PACKAGES_DIR",
+            "BUILDSYS_SOURCES_DIR",
+            "BUILDSYS_VARIANT",
```
This will trigger a rebuild of all packages whenever the variant changes. We need to continue using the selective logic in main.rs instead.
```suggestion
```

bcressey commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,143 @@
+            "BUILDSYS_PACKAGES_DIR",
+            "BUILDSYS_SOURCES_DIR",
+            "BUILDSYS_VARIANT",
+            "CARGO_PKG_NAME",
```
We don't track this for changes today.
```suggestion
```

bcressey commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,143 @@
+            "BUILDSYS_NAME",
+            "BUILDSYS_PRETTY_NAME",
+            "BUILDSYS_VARIANT",
+            "BUILDSYS_VERSION_BUILD",
+            "BUILDSYS_VERSION_IMAGE",
```
This list should have the other variables that only drive variant rebuilds.
```suggestion
            "BUILDSYS_VERSION_BUILD",
            "BUILDSYS_VERSION_IMAGE",
            "BUILDSYS_VERSION_FULL",
            "BUILDSYS_TIMESTAMP",
```

bcressey commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,143 @@
+    #[arg(long, env = "BUILDSYS_VERSION_FULL")]
+    pub(crate) version_full: String,
+
+    #[arg(long, env = "CARGO_MANIFEST_DIR")]
+    pub(crate) cargo_manifest_dir: PathBuf,
```
I'm not as excited about converting cargo's own environment variables into arguments since the environment is the canonical way to pass this information.

webern replied:
My preference would be to include it here anyway for two reasons:
1) Readability: self-documenting code
2) If I want to write buildsys tests, I may call it directly instead of invoking it from a build.rs script and using args might be helpful for that

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/b535980ed80a54ce9aaa803de228e4120a3ad096..a1eae7b2165900eca3e07e754c7c0b5a09469b89

Fixup the cargo:rerun-if-env-changed list to make it the same as it used to be.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/a1eae7b2165900eca3e07e754c7c0b5a09469b89..c71f56ece7a6e1df398e5552a96c756b34a1e51f

Make the version a parameter of LookasideCache

webern committed:
buildsys: refactor builder.rs

Refactor the code, primarily in builder.rs. The goal of the refactor is
to segregate data collection and docker argument creation from program
flow logic.

webern commented:
I've gotten myself into a much bigger refactor of builder.rs so converted this to a draft.

webern commented:
@bcressey this is ready for you to re-review. The refactor is rather substantial, but I think it is an improvement. My goal was to separate the concerns of a) gathering data from the environment, b) manufacturing and constructing args for docker, and c) the program flow of calling docker. Previously all of those things were happening in the same code-paths. I think having them in separate code paths will make it easier to hack on this for kits.

webern commented on tools/buildsys/src/builder.rs:
```diff
@@ -86,269 +85,355 @@ lazy_static! {
+    target: String,
+    tag: String,
+    root_dir: PathBuf,
+    // build_dir: PathBuf,
```
I fixed this.
```suggestion
```

webern committed:
buildsys: use clap for env vars

Required input in the form of environment variables was sprinkled
throughout the code. Here we aggregate the inputs into a CLI interface
using Clap, while still allowing all to specified via environment
variables.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/9a89bf58869592d3f451af59a5cb35b5c81a9329..be37869a03eda322ed07e527aacd978a5b0a8bb9

Slightly painful rebase.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/be37869a03eda322ed07e527aacd978a5b0a8bb9..8b857d98fb2aed710b662049273591b2e2888596

Clippy fix.

bcressey commented on tools/buildsys/src/builder.rs:
```diff
@@ -86,269 +85,355 @@ lazy_static! {
+                    .image_features()
+                    .map(|set| set.iter().map(|&x| x.to_owned()).collect()),
+                package,
+                publish_repo: args.publish_repo,
```
I opened #151 to correct this since it felt out of place here.

bcressey commented on tools/buildsys/src/builder.rs:
```diff
@@ -86,269 +85,355 @@ lazy_static! {
-    docker(&rm, Retry::No)?;
-
-    // Clean up our image now that we're done.
-    docker(&rmi, Retry::No)?;
-
-    // Copy artifacts to the expected directory and write markers to track them.
-    copy_build_files(&build_dir, output_dir)?;
+            context = self.context.display(),
+            dockerfile = self.dockerfile.display(),
+            target = self.target,
+            tag = self.tag,
```
nit: the spacing on these lines seems off?

webern replied:
> nit: the spacing on these lines seems off?

Yeah fmt is no help here. I've now aligned the multiline string line starts with the variable list below it.

bcressey commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,214 @@
+use std::path::PathBuf;
+use url::Url;
+
+/// A list of environment variables and the type of build that should be rerun if that environment
+/// variable changes. The build type is represented with bit flags so that we can easily list
+/// multiple build types for a single variable. See `[BuildType]` and `[rerun_for_envs]` below to
+/// see how this list is used.
+const REBUILD_VARS: [(&str, u8); 13] = [
+    ("BUILDSYS_ARCH", PACKAGE | VARIANT),
```
❤️ 

This is a nice way to organize something that was very ad-hoc and organic before.

bcressey commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,214 @@
+const REBUILD_VARS: [(&str, u8); 13] = [
+    ("BUILDSYS_ARCH", PACKAGE | VARIANT),
+    ("BUILDSYS_NAME", VARIANT),
+    ("BUILDSYS_OUTPUT_DIR", VARIANT),
+    ("BUILDSYS_PACKAGES_DIR", PACKAGE),
```
nit: it feels like a (pre-existing) mistake to not track `BUILDSYS_SOURCES_DIR` if we are tracking `BUILDSYS_PACKAGES_DIR` - if we override the build to point to a completely different sources tree then that should invalidate any builds that used those sources.

webern replied:
> nit: it feels like a (pre-existing) mistake to not track `BUILDSYS_SOURCES_DIR` if we are tracking `BUILDSYS_PACKAGES_DIR` - if we override the build to point to a completely different sources tree then that should invalidate any builds that used those sources.

I'm going to defer this because I'm concerned that the symlink change when changing variant could trigger something. I'll track this as a GitHub issue and play around with adding it. https://github.com/bottlerocket-os/twoliter/issues/152

Edit: no that's dumb. I can safely add this.

Edit, edit: Actually I'm still squeamish... deferring.

bcressey commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,214 @@
+    #[arg(long, env = "BUILDSYS_PACKAGES_DIR")]
+    pub(crate) packages_dir: PathBuf,
+
+    #[arg(long, env = "BUILDSYS_VARIANT")]
+    pub(crate) variant: String,
```
To check my understanding - we don't want this in `Common` because kits won't need this variable?

webern replied:
> To check my understanding - we don't want this in `Common` because kits won't need this variable?

That's the idea, yeah. And I think our goal should be that package builds also don't need to know the variant name.

ecpullen commented on tools/buildsys/src/args.rs:
```diff
@@ -0,0 +1,214 @@
+    #[arg(long, env = "BUILDSYS_PACKAGES_DIR")]
+    pub(crate) packages_dir: PathBuf,
+
+    #[arg(long, env = "BUILDSYS_VARIANT")]
```
This goes away once we no longer have conditional compilation right?

webern replied:
> This goes away once we no longer have conditional compilation right?

That should be the goal for packages, yes.

ecpullen reviewed (approved):
LGTM

yeazelm commented on tools/buildsys/src/builder.rs:
```diff
@@ -86,269 +85,355 @@ lazy_static! {
-    let mut build = format!(
-        "build . \
+        let mut build = format!(
+            "build {context} \
```
nit: this intending looks strange to me?

webern replied:
> nit: this intending looks strange to me?

Yeah fmt is no help here. I've now aligned the multiline string line starts with the variable list below it.

yeazelm commented on tools/buildsys/src/builder.rs:
```diff
@@ -358,6 +443,8 @@ fn docker(args: &[String], retry: Retry) -> Result<Output> {
         retry_messages = messages;
     }
 
+    println!("Running command:\ndocker {}", args.join(" /\n"));
```
This is only one place out of two we `println!()` in this file. Is that intentional? I think so but wanted to ensure this isn't a hold over from debugging and this output is doing something useful for logs/output for the user?

webern replied:
> This is only one place out of two we println!() in this file. Is that intentional? I think so but wanted to ensure this isn't a hold over from debugging and this output is doing something useful for logs/output for the user?

Good catch. Fixed.
